src/sb_handle/temp_memory_pool.hpp

```diff
@@ -0,0 +1,118 @@
+#ifndef TEMP_MEMORY_POOL_HPP
+#define TEMP_MEMORY_POOL_HPP
+
+#include "portblas_helper.h"
+
+namespace blas {
+template <typename value_t>
+typename helper::AllocHelper<value_t, helper::AllocType::buffer>::type
+Temp_Mem_Pool::acquire_buff_mem(size_t size) {
+  const size_t pad = sizeof(double) / sizeof(value_t);
+  // Adjust the requested size in order to reinterpret for double's
+  size += (pad - size % pad);
+  const size_t byteSize = size * sizeof(value_t);
+  temp_buffer_map_mutex_.lock();  // lock
+  auto found = temp_buffer_map_.lower_bound(byteSize);
+  if (found != temp_buffer_map_.end()) {
+    cl::sycl::buffer<temp_buffer_map_t::mapped_type::value_type, 1> buff =
+        found->second;
+    temp_buffer_map_tot_byte_size_ -= found->first;
+    temp_buffer_map_.erase(found);
+    temp_buffer_map_mutex_.unlock();  // unlock
+    return blas::BufferIterator<value_t>{buff.reinterpret<value_t>(
+        cl::sycl::range<1>(buff.byte_size() / sizeof(value_t)))};
+  } else {
+    temp_buffer_map_mutex_.unlock();  // unlock
+#ifdef VERBOSE
+    std::cout << "Create a temporary buffer of " << byteSize << " bytes."
+              << std::endl;
+#endif
+    return make_sycl_iterator_buffer<value_t>(size);
+  }
+}
+
+template <typename container_t>
+void Temp_Mem_Pool::release_buff_mem_(const container_t& mem) {
+  const size_t byteSize = mem.get_buffer().byte_size();
+  auto rebuff =
+      mem.get_buffer()
+          .template reinterpret<temp_buffer_map_t::mapped_type::value_type>(
+              cl::sycl::range<1>(
+                  byteSize /
+                  sizeof(temp_buffer_map_t::mapped_type::value_type)));
+  temp_buffer_map_mutex_.lock();  // lock
+  if (temp_buffer_map_tot_byte_size_ + byteSize <= max_size_temp_mem_) {
+    temp_buffer_map_tot_byte_size_ += byteSize;
+    temp_buffer_map_.emplace(byteSize, rebuff);
+  }
+  temp_buffer_map_mutex_.unlock();  // unlock
+}
+
+template <typename container_t>
+typename Temp_Mem_Pool::event_t Temp_Mem_Pool::release_buff_mem(
```

Comment:
Could you please explain why we need a host task for releasing buffers here?

Reply:
Thanks for your review, @muhammad-tanvir-1211.

We  call the `release_buff_mem_`  via a `host_task` because we need to insert the buffer within the map just when the involved computation is finished. Otherwise, a buffer still in use could be passed to another operation creating unnecessary dependencies between unrelated operations.